### ISSUE TITLE
"guest" time is double-counted in "user" time

### ISSUE THREAD

**Issue by candlerb** [firehol/netdata#729]
TL;DR: the `proc(5)` man page says:

~~~
              (14) utime  %lu
                        Amount of time that this process has been scheduled
                        in user mode, measured in clock ticks (divide by
                        sysconf(_SC_CLK_TCK)).  This includes guest time,
                        guest_time (time spent running a virtual CPU, see
                        below), so that applications that are not aware of
                        the guest time field do not lose that time from
                        their calculations.
~~~

Therefore, to find user time separately from guest time, you have to subtract the guest time.

But it appears netdata is not doing this. Here's how I reproduce the problem:

* Base machine running Ubuntu 14.04 with linux-generic-lts-xenial kernel (4.4.0). This machine has 4 real cores and hyperthreading, so appears as 8 cores.
* Inside this is a KVM VM running Ubuntu 16.04 (also 4.4.0 kernel). I have configured this VM to see exactly the same set of cores (details below).
* Netdata is installed both in the VM and in the outer host

Now I run an infinite loop in the guest:

~~~
# while [ 1 ]; do a=123; done
~~~

Inside the VM, this is what I see:

![image](https://cloud.githubusercontent.com/assets/44789/17148878/bbad5a78-5360-11e6-83f5-51941561361a.png)

That's reasonable: one out of 8 cores fully utilised, 12.5% utilised. And 'top' shows:

~~~
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
13312 root      20   0   24848   5020   3204 R 100.0  0.0   5:10.40 bash
~~~

But in the outer host, this is what I see:

![image](https://cloud.githubusercontent.com/assets/44789/17149025/5e420536-5361-11e6-9909-9436598b85c8.png)

It seems that both "user" and "guest" are accounting for 11.5% each.

'top' shows:

~~~
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
29106 libvirt+  20   0 17.081g 924272  16592 S 102.2  5.7   6:32.48 qemu-system-x86
~~~

According to the manpage quoted above the qemu/kvm time is included in both "user" and "guest" figures; so I believe that when the "guest" figure is present, netdata should be subtracting it from "user". See also [this stackoverflow answer](http://stackoverflow.com/questions/23367857/accurate-calculation-of-cpu-usage-given-in-percentage-in-linux#answer-23376195)

----

libvirt config:

~~~
  <vcpu placement='static'>8</vcpu>
...
  <cpu mode='host-passthrough'>
    <topology sockets='1' cores='4' threads='2'/>
  </cpu>
~~~

This results in the following flags on the qemu-system-x86_64 command line:

~~~
-smp 8,sockets=1,cores=4,threads=2
~~~

**Comment by candlerb** [comment]
Additional info: if I look at the "CPUs" graphs (cpu.cpu0 to cpu.cpu7), then in the inner VM I see one core at 100% utilisation and the others idle.

In the outer host, I also see one core at 100% utilisation, but it appears as 50% user and 50% guest.

![image](https://cloud.githubusercontent.com/assets/44789/17149312/a3e7c70a-5362-11e6-9f69-7cbf0d3bb769.png)

So I think there's a similar problem here.

However if I look at `apps.cpu`, `apps.cpu_user`, `apps.cpu_system` then the inner and outer hosts are consistent: both show a single app using 100% of user time (this is "ssh" in the inner VM and "other" in the outer host)

**Comment by candlerb** [comment]
... and here is what started me down this road.

In the inner VM I am running something which uses some background CPU (dynamips + dynagen). Once dynamips has finished starting up, looking at the per-core stats inside the VM they bob around 4%, occasionally peaking at 10%

![image](https://cloud.githubusercontent.com/assets/44789/17149851/01c33e5c-5365-11e6-8d96-9bedec46d7df.png)

But in the outer host it seems substantially more:

![image](https://cloud.githubusercontent.com/assets/44789/17149865/0fa9dd28-5365-11e6-933d-e952970f8ec7.png)

Part of that can be explained by the user/guest doubling, but I don't think it fully does. So once this is resolved, I will be digging further.

Maybe I'm looking at virtualisation overhead; and/or this might reflect that the outer host actually has only 4 real CPUs, which are timeslicing with hyperthreading (and therefore each virtual core has lower capacity than a real one)

**Comment by ktsaou** [comment]
nice finding! thanks for reporting it!

- [ ] However, for the CPU charts netdata reads this info from `/proc/stat`. On the same man page I don't see anything relevant. **I can fix it though in netdata**.

- [ ] For the Applications, Users and User Groups charts netdata uses the info you mention, but it ignores guest values (so they appear right). **I can fix that too**.

In `/proc/stat` I see there is also `guest_nice` reported: https://github.com/firehol/netdata/blob/8abfd74e85d19903559ca65da92796aaa395810d/src/proc_stat.c#L68-L78

Can you detect if this is included in `nice` or `user`?

**Comment by candlerb** [comment]
Minor note: the "guest" time is almost always less than or equal to the "user" time. But occasionally I have caught a time point where this is not true. Here is one:

![image](https://cloud.githubusercontent.com/assets/44789/17151158/81827f0e-536a-11e6-85ae-63bf961dd038.png)

Hence some defensive clipping (to avoid negative user time) may be required.

**Comment by candlerb** [comment]
> However, for the CPU charts netdata reads this info from /proc/stat. On the same man page I don't see anything relevant. 

True, it's not clear at this point. It says only:

~~~
                     guest (since Linux 2.6.24)
                            (9)  Time  spent  running  a virtual CPU for guest
                            operating systems under the control of  the  Linux
                            kernel.

                     guest_nice (since Linux 2.6.33)
                            (10) Time spent running a niced guest (virtual CPU
                            for guest operating systems under the  control  of
                            the Linux kernel).
~~~

However I found a [stackoverflow answer](http://stackoverflow.com/questions/23367857/accurate-calculation-of-cpu-usage-given-in-percentage-in-linux#answer-23376195) which points to the source code of [htop](https://github.com/hishamhm/htop/blob/e0209da88faf3b390d71ff174065abd407abfdfd/ProcessList.c#L940) which does parse /proc/cpuinfo, and it does perform the subtraction of guest and guest_nice from user and user_nice.

> In /proc/stat I see there is also guest_nice reported:  ... Can you detect if this is included in nice or user?

In the htop source it indicates that guest_nice is included within nice:

~~~
      // Guest time is already accounted in usertime
      usertime = usertime - guest;
      nicetime = nicetime - guestnice;
~~~

Strangely though, I tried doing a "renice" on a running qemu-kvm process and didn't see any difference in the graphs: still seeing user/guest time around 5%, and nice <1%.

~~~
$ head -1 /proc/stat; sleep 10; head -1 /proc/stat
cpu  954616 339036 1107547 68224360 15406 0 14710 0 757929 0
cpu  955325 339049 1107652 68231542 15408 0 14712 0 758630 0
~~~

The second and last numbers are nice and guestnice respectively, and you can see not much is in either. But 'top' does show that niceness 10 has been applied:

~~~
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
19123 libvirt+  30  10 16.581g 2.898g  16516 S  62.8 18.6  23:55.06 qemu-syste+
17905 root      20   0 5901780 230420   9332 S   2.7  1.4  33:55.37 java
   51 root      25   5       0      0      0 S   2.3  0.0   8:17.74 ksmd
13037 netdata   39  19  286512  19704   7520 S   1.3  0.1   0:37.01 python
13039 netdata   39  19   16204   3132   1920 S   1.3  0.0   0:35.36 apps.plugin
16204 mongodb   20   0  382064  17436   2692 S   1.0  0.1  11:40.61 mongod
13028 netdata   39  19 1078332  44824   2068 S   0.7  0.3   0:23.49 netdata
~~~

Very odd.

**Comment by ktsaou** [comment]
Can you try my repo?

```sh
git clone https://github.com/ktsaou/netdata
```

I did all the above. A new chart to report guest time for applications, users and user groups has been added.

**Comment by ktsaou** [comment]
Pay some attention to clipping. It is not so straight forward because the values collected are incremental (i.e. the are always getting bigger). I think it will work by itself (it will consider the values as reset, but have a look).

**Comment by candlerb** [comment]
Thank you!

The overall CPU graph is right now, as are the individual CPU graphs: e.g.

![image](https://cloud.githubusercontent.com/assets/44789/17153516/fc3485bc-5374-11e6-906f-d926ce66dec4.png)

I highlighted a point where the 'user' value is negative. You can see from looking along the horizontal axis that this happens perhaps a couple of times a minute.

> A new chart to report guest time for applications, users and user groups has been added.

Sorry but I can't find this. Under "Applications" I still see "apps.cpu", "apps.cpu_user", "apps.cpu_system": ditto under "users" and "groups". Should I look somewhere else, or does the config need to be updated?

I just re-ran `./netdata-installer.sh --install /opt`

**Comment by ktsaou** [comment]
Refresh the dashboard. Under Applications in subsection "cpu" there should be 4 chart now. The 4th is the guest. The same should happen for Users and User Groups.

**Comment by ktsaou** [comment]
I am trying to decide what I should do with the negative values. I think I should not do anything. This is purely a kernel issue that should not be hidden. I don't know. I don't like it, but I think it is not right to fix it in netdata either.

**Comment by ktsaou** [comment]
Can you check if the negative values appear on the per-core charts too? It might be an issue on the way the kernel sums them to find the total cpu of the system. If this is the case (i.e. the per core charts do not have any negative values), it might be wise for netdata to sum them to find the system total, instead of using the kernel supplied values for it.

**Comment by candlerb** [comment]
Ah OK I see them after refresh. But all of them are showing zero for guest. e.g.

![image](https://cloud.githubusercontent.com/assets/44789/17156405/9808fade-5381-11e6-8b08-4eda142a2973.png)

Note: I did add to `/opt/netdata/etc/netdata/app_groups.conf`

~~~
qemu: qemu-system-*
~~~

But if I take that out, all the app/user/group CPU time is accounted as "other" instead. Still zero for "guest".

**Comment by candlerb** [comment]
> Can you check if the negative values appear on the per-core charts too?

They do - it was actually one of the per-core charts I showed previously (cpu.cpu0). The total CPU does also occasionally show a small negative blip:

![image](https://cloud.githubusercontent.com/assets/44789/17156568/6d8d5510-5382-11e6-8cc7-a23fa43e4c8b.png)

I think you're right to leave this as is, to expose some underlying synchronization problem. (It should be possible to replicate this simply by cat'ing /proc/stat and looking for instances where the guest column increments by a larger number that the user column)

**Comment by ktsaou** [comment]
> But if I take that out, all the app/user/group CPU time is accounted as "other" instead. Still zero for "guest".

Could you please run this for the PID of qemu?

```sh
p=PID_OF_QUEMU
cat /proc/$p/stat | cut -d ' ' -f 43-44
```

According to the man page these 2 values are the `guest` and `cguest` usage.

**Comment by candlerb** [comment]
Result:

~~~
2685608 0
~~~

Picking out utime, cutime, guest_time, cguest_time:

~~~
$ cat /proc/19123/stat | cut -d' ' -f 14,16,43,44
2829491 0 2706044 0
~~~

Clearly guest time *is* present, and accounts for the majority of the user time.

**Comment by candlerb** [comment]
Regarding the negative values: I did the following capture:

~~~
for i in {1..60}; do head -1 /proc/stat >>stat.out; sleep 1; done
~~~

And a quick python prog to scan the results:

~~~
import sys
ug1 = 0
g1 = 0
line1 = ""
for line2 in sys.stdin:
    cols = line2.split()
    ug2, g2 = int(cols[1]), int(cols[9])
    ug, g = ug2 - ug1, g2 - g1
    u = ug - g
    if u < 0:
        print "***"
        print line1
        print line2
        print "*** g=%d u=%d ***" % (g,u)
    line1, g1, ug1 = line2, g2, ug2
~~~

Result: in one minute I found two instances where utime incremented by less than (utime+guest)

~~~
***
cpu  3644262 407756 1810922 100557916 21583 0 20305 0 3388311 0

cpu  3644306 407756 1810928 100558638 21583 0 20305 0 3388356 0

*** g=45 u=-1 ***
***
cpu  3646033 407806 1811197 100583592 21596 0 20308 0 3389980 0

cpu  3646084 407808 1811203 100584304 21596 0 20308 0 3390032 0

*** g=52 u=-1 ***
~~~

So this is normal behaviour for this host which is running a VM and doing not much else.

Whilst I don't mind seeing the negative values too much, I think it would also be reasonable to increase the utime value read; this would in effect 'borrow' the negative time from the next time tick.

**Comment by ktsaou** [comment]
The apps, users, groups guest time chart should be ok now, including negative values.

I'll try to find a solution for the negative values of the system CPU charts.

**Comment by ktsaou** [comment]
I pushed another commit. So, if you tried it already, please do it again.

**Comment by candlerb** [comment]
Thanks, that looks better:

![image](https://cloud.githubusercontent.com/assets/44789/17171491/d45657a4-53e8-11e6-9f58-24dbc8302219.png)

The time spent in "user" has dropped, and the time spent in "guest" has gone up accordingly.

(Aside: the guest graph Y axis does not extend down to zero, perhaps it should? That's a decision about how auto-scaling should work, but for CPU% I think it would be less confusing to always include zero)

I note also that there is a significant amount of time in "system" relating to the qemu process. That's probably correct: i.e. linux separates user and guest user time, but kvm time spent in system is just system time.

But this means it may not actually be all that useful to show app guest time in a separate graph. You may as well see qemu user time in the user graph, and qemu system time in the system graph. What do you think?

Similarly in the group graph:

![image](https://cloud.githubusercontent.com/assets/44789/17171606/5befe270-53e9-11e6-9a8c-6a03347efb45.png)

You can see kvm group in the 'guest' graph and in the 'system' graph.

**Comment by candlerb** [comment]
I pulled in your rrd changes from ddac4532. Over 5 minutes I have not observed any negative downticks in the cpu.cpuX graphs. This looks good to me!

![image](https://cloud.githubusercontent.com/assets/44789/17172026/29651972-53eb-11e6-98d8-20bf060eef61.png)

**Comment by ktsaou** [comment]
I added the `apps.plugin` command line options:

- `with-guest` to enable guest charts (separate guest charts in apps, users, groups)
- `without-guest` to disable guest charts (guest will be reported as `user` in apps, users, groups)

To set them, edit netdata.conf and find this section:

```
[plugin:apps]
	# update every = 1
	# command options = 
```

uncomment `command options` and set it like this:

```
[plugin:apps]
	# update every = 1
	command options = with-guest
```

Could you please test both options and report back?
The default is now `without-guest`.

**Comment by candlerb** [comment]
There is an anomoly here, which may or may not be to do with Linux's own accounting.

Look at the total CPU graph: you can see user time is almost zero and this is consistent. Most stuff is in guest time, with a bit of system time.

![image](https://cloud.githubusercontent.com/assets/44789/17172227/fc2cadac-53eb-11e6-99f3-014acf07aa74.png)

However, looking at "apps":

![image](https://cloud.githubusercontent.com/assets/44789/17172255/156ee3ac-53ec-11e6-8fdc-3dc97f7fcb20.png)

And "user groups":

![image](https://cloud.githubusercontent.com/assets/44789/17172281/30abdf1c-53ec-11e6-92fe-7892b13f86af.png)

Firstly, there is a signficiant portion of "user" time as well as "guest". And then, a chunk of this time moves away from user and into guest!

This is weird. Given the total CPU showing minimal user time, I was expecting no app user time.

What is the underlying source of this data? Is it `/proc/<pid>/stat` ? Testing:

~~~
for i in {1..60}; do cat /proc/19123/stat | cut -d' ' -f14,43; sleep 1; done >stats.out
~~~

Results:

~~~
3455660 3299633
3455724 3299668
3455793 3299725
3455868 3299782
3455934 3299833
3456000 3299883
3456065 3299927
3456135 3299980
3456204 3300027
3456268 3300068
...
~~~

Iindeed it looks like the process is accruing both user and guest time:

~~~
ug=64 g=35 u=29
ug=69 g=57 u=12
ug=75 g=57 u=18
ug=66 g=51 u=15
ug=66 g=50 u=16
ug=65 g=44 u=21
ug=70 g=53 u=17
ug=69 g=47 u=22
ug=64 g=41 u=23
ug=64 g=42 u=22
...
~~~

Although the *total* figures suggest the vast majority historically has been guest (95%)

Oh well, I can't ask netdata to show anything other than what the kernel is telling it :-)

**Comment by candlerb** [comment]
Re latest version: yes, 'guest' graph vanishes, and comes back with "command options = with-guest"

**Comment by ktsaou** [comment]
> latest version: yes, 'guest' graph vanishes, and comes back with "command options = with-guest"

ok, and I guess the charts show the proper values.

Regarding the files, apps, users and user groups come from `/proc/pid/stat`, while system figures come from `/proc/stat`.

The reported values do not always agree. I think the kernel accounting is probably rounding values under certain conditions.

The only calculation performed by apps.plugin is to scale the values read by `/proc/pid/stat` files, so that their total is always equal or less to `/proc/stat` (yes, values collected from processes are in a few cases higher than the total of the system, check #644 - and netdata works around this issue by calculating a scaling factor for the cpu utilisation reported in apps, users and user groups - however this scaling cannot affect their relationship, i.e. user, system and guest time are all scaled by the same factor).

**Comment by candlerb** [comment]
> The reported values do not always agree. I think the kernel accounting is probably rounding values under certain conditions.

I think what I see is more severe than simply rounding: /proc/stat says that there is virtually zero CPU in "user" (i.e. it's all "guest"), whilst apps/user/group are showing significant proportion of time in "user".

In fact, it's not even internally consistent within apps:

![image](https://cloud.githubusercontent.com/assets/44789/17178047/d273e550-540b-11e6-99dc-b5eb741b251b.png)

Note that apps.cpu shows a peak at 200%, and I would expect apps.cpu = apps.cpu_user + apps.cpu_system, but the total of those is much less.

How is apps.cpu calculated? Presumably it's not simply the sum of apps.cpu_user and apps.cpu_system?

(I can move this to separate issue if you prefer)

**Comment by candlerb** [comment]
I have a suspicion the problem is around here:

~~~
void send_collected_data_to_netdata(struct target *root, const char *type, unsigned long long usec
) {
        struct target *w;

        send_BEGIN(type, "cpu", usec);
        for (w = root; w ; w = w->next) {
                if(unlikely(w->exposed))
                        send_SET(w->name, (unsigned long long)(w->utime * utime_fix_ratio) + (unsigned long long)(w->stime * stime_fix_ratio) + (unsigned long long)(w->gtime * gtime_fix_ratio) + (include_exited_childs?((unsigned long long)(w->cutime * cutime_fix_ratio) + (unsigned long long)(w->cstime * cstime_fix_ratio) + (unsigned long long)(w->cgtime * cgtime_fix_ratio)):0ULL));
        }
~~~

Note that gtime is added here into the total, even if show_guest_time is unset. That would explain why the "app.cpu" graph does not equal the sum of "app.cpu_user" and "app.cpu_system".

In this mode, guest time has not been subtracted from user time, and so it has been double-counted. Simple fix is to force gtime/cgtime to zero when not showing guest graphs?

**Comment by candlerb** [comment]
Possibly similar issue here:

~~~
                if(global_utime + global_stime + global_gtime > utime + cutime + stime + cstime + gtime + cgtime) {
~~~

i.e. global_gtime may be non-zero, but is not subtracted from `global_utime` unless `enable_guest_charts` is true; same possibly applies to `[c]gtime` which is sum of `w->[c]gtime`

**Comment by ktsaou** [comment]
You are right.
I just pushed another commit that does not collect guest usage when running `without-guest`.
Could you please test it?

**Comment by candlerb** [comment]
Yes that's much better: apps/groups/users.cpu now matches cpu_user + cpu_system

**Comment by ktsaou** [comment]
ok, perfect!

I leave it unmerged until tomorrow. If you find something please report it here. If nothing abnormal if found until tomorrow, I'll merge it.

Thanks!

**Comment by ktsaou** [comment]
merged it.
Thank you for reporting this and helping me out setle it.
